### PR TITLE
Document import paths, improve tests in process of understanding code

### PR DESCRIPTION
**Pull request by hornc**
### Description

* Document UI for adding books. openlibrary/plugins/upstream/addbook.py
* Document catalog/addbook import path for MARC / non-MARC API, and ISBN imports 
* Add tests for many methods used in import to confirm expected data inputs and outputs and results
* split out existing tests under catalog/ into their own files
* Added deprecation warnings to a range of duplicate methods under catalog to ensure future work is done on the versions that are actually being used
* Changes the test mock_site in openlibrary/mocks/mock_infobase.py to support queries using`isbn_` which I confirmed is the behaviour of the live site. This test object has its own tests, so I have added tests to verify it works as intended. Without it, import code under test was not behaving like prod.

### Technical

This PR adds a bunch of python deprecation warnings to /catalog/amazon and /catalog/merge code that I _believe_ is not used. If they are used anywhere, the usage needs to be unified. Hopefully the warnings will appear in logs and be noticeable without breaking anything.

I have marked them as deprecated because I'm not 100% sure they aren't used elsewhere, and also, unfortunately, I'm not completely sure the currently used live versions of the methods are better than the deprecated ones. There is still some investigation to complete. For example, the currently live used version of `compare_authors()`  in merge_marc does not make use of  `by_statement` whereas the _two_ other (now deprecated) versions in amazon.py and merge.py do. This looks like a loss of functionality, but I can't tell if there isn't a good reason for it now.

TL;DR the deprecation warnings are on duplicate methods that aren't the main ones used by the current live import process, it doesn't necessarily mean they are inferior. It represents the current state, not necessarily the ideal. Fun times :)

### Testing

I have only made minimal refactor changes to the code, the vast bulk of the changes are docstrings and adding _new_ tests of existing functionality to help understand which code is used by the Live site and how data is handled going through.

There is more work I could do, the overall merge / matching algorithm still does not have proper test coverage, the last commit was a start in that direction. Some of the tests I have split out should eventually be removed when that deprecated functionality has been proven to work on the latest code. It feels like a focused task for a new PR. I need to make some progress on some of the features and fixes (which this PR enables) e.g. #2311

### CODE REVIEW

File: openlibrary/plugins/upstream/addbook.py [hunk 1]
```diff
@@ -173,16 +184,24 @@ def POST(self):
...
+        """
+        Tries to find an edition or a work or multiple works that match the given input data.
 
         Case#1: No match. None is returned.
         Case#2: Work match but not editon. Work is returned.
         Case#3: Work match and edition match. Edition is returned
         Case#3A: Work match and multiple edition match. List of works is returned
         Case#4: Multiple work match. List of works is returned.
+
+        :param DocSaveHelper saveutil:
```
Comment (hornc) [inline openlibrary/plugins/upstream/addbook.py]: `saveutil` should not be passed to `find_matches()`, which apparently can create an author and return `None`. The author creation should not occur inside a `find` method... move the author creation to `addbook.POST()` and remove `saveutil` from `find_matches()`

File: openlibrary/plugins/upstream/addbook.py [hunk 2]
```diff
@@ -173,16 +184,24 @@ def POST(self):
...
 
     def find_matches(self, saveutil, i):
-        """Tries to find an edition or a work or multiple works that match the given input data.
+        """
+        Tries to find an edition or a work or multiple works that match the given input data.
 
         Case#1: No match. None is returned.
         Case#2: Work match but not editon. Work is returned.
         Case#3: Work match and edition match. Edition is returned
         Case#3A: Work match and multiple edition match. List of works is returned
```
Comment (hornc) [inline openlibrary/plugins/upstream/addbook.py]: `Case#3A: Work match and multiple edition match. List of works is returned` does not make sense, what is this case, and what does it return?

File: openlibrary/catalog/add_book/test_add_book.py [hunk 3]
```diff
@@ -56,6 +62,21 @@ def test_build_query(add_languages):
...
+    assert isinstance(result, list)
+    assert '9780190906764' in result
+    assert '0190906766' in result
+    assert len(result) == 2
+
+def test_editions_matched(mock_site):
+    rec = {'title': 'test', 'isbn_13': ['9780190906764'], 'isbn_10': ['0190906766']}
+    isbns = isbns_from_record(rec)
+    result = editions_matched(rec, 'isbn_', isbns)
+    #TODO: does a thing search with key = 'isbn_' work?
```
Comment (hornc) [inline openlibrary/catalog/add_book/test_add_book.py]: Yes, **confirmed** key `isbn_` will return a matching record on either `isbn_10` or `isbn_13` metadata fields!

Comment (hornc) [reply openlibrary/catalog/add_book/test_add_book.py]: unfortunately I can't get the test to pass with a match using `isbn_`, something is different. Needs further testing

Comment (hornc) [reply openlibrary/catalog/add_book/test_add_book.py]: Looks like this works on the live site, but not with the test `mock_site` see note at https://github.com/internetarchive/openlibrary/blob/75452cf6f0e3c23e1030395c93e232686ebaa417/openlibrary/records/tests/test_functions.py#L232

Comment (hornc) [reply openlibrary/catalog/add_book/test_add_book.py]: I have improved `mock_site` so it behaves like the live site in terms of supporting queries by `isbn_` -- not finding any matches on `isbn_` was skipping a whole lot of code paths in and after `openlibrary.catalog.add_book.editions_matched()` which was always returning `[]`

File: openlibrary/catalog/merge/amazon.py [hunk 4]
```diff
@@ -125,9 +115,9 @@ def compare_authors(amazon, marc):
         return ('authors', 'field missing from one record', -25)
 
     for name in amazon['authors']:
-        if 'authors' in marc and match_name(name, marc['authors'][0]['name']):
+        if 'authors' in marc and match_name(name[0], marc['authors'][0]['name']):
```
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/catalog/merge/amazon.py]: I have reverted these changes, and modified the existing tests so they describe the code as written here. Unfortunately the 'duplicate' method `openlibrary.catalog.merge.merge_marc.compare_authors()` is slightly different. I'm assuming `merge_marc` is newer, but this code looks like it does more with `by_statement`s

File: openlibrary/catalog/merge/amazon.py [hunk 5]
```diff
@@ -98,13 +89,12 @@ def compare_date(e1, e2):
         return ('publish_date', 'mismatch', -250)
 
 def compare_isbn10(e1, e2):
-    if len(e1['isbn']) == 0 or len(e2['isbn']) == 0:
+    if not (e1.get('isbn_10') and e2.get('isbn_10')):
```
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/catalog/merge/amazon.py]: reverted, and using existing tests to test the code as is is currently.

File: openlibrary/catalog/merge/test_merge_marc.py [hunk 6]
```diff
@@ -33,30 +135,17 @@ def test_merge():
                    u'eli whitney and the birth of american technology']}
 
     assert compare_authors(bpl, lc) == ('authors', 'exact match', 125)
-    threshold = 735
+    threshold = 875
```
Comment (mekarpeles) [inline openlibrary/catalog/merge/test_merge_marc.py]: Is this a magic #? How is it chosen
